### PR TITLE
In flood_fill deprecate inplace and use in_place instead

### PR DESCRIPTION
**Pull request by robroooh**
## Description

enhancement from https://github.com/scikit-image/scikit-image/issues/4248. adding deprecation warning for flood_fill param.

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

**Comment by hmaarrfk** [comment]
This would be included in 0.17, and removed in 0.19

**Comment by sciunto** [comment]
You need to add in the function signature ", *, in_place=False", turns default for "inplace" to None, document the new option, update the former one and state that it's deprecated, handle both arguments at the top of the function to use in_place everywhere else in the function.

You also need to check for other function calls in the library, especially in the documentation :)

**Comment by robroooh** [comment]
updated the pr now :smiley: Please give me feedback on what could be improved :+1:

File: skimage/morphology/tests/test_flood_fill.py [hunk 1]
```diff
@@ -93,7 +93,7 @@ def test_inplace_noncontiguous():
     # Transpose is noncontiguous
     image2 = image[::2, ::2]
 
-    flood_fill(image2, (0, 0), 5, inplace=True)
+    flood_fill(image2, (0, 0), 5, in_place=True)
```
Comment (hmaarrfk) [inline skimage/morphology/tests/test_flood_fill.py]: Can you duplicate one of these tests to ensure that the old parameter still works?

That test can then be easily removed when it comes time to finally rice the behavior. There is also a way to enforce q check on the working being issued.

Comment (robroooh) [reply skimage/morphology/tests/test_flood_fill.py]: That's a very nice input! Thanks :+1:

File: skimage/morphology/_flood_fill.py [hunk 2]
```diff
@@ -100,10 +105,15 @@ def flood_fill(image, seed_point, new_value, *, selem=None, connectivity=None,
            [5, 5, 5, 5, 2, 2, 5],
            [5, 5, 5, 5, 5, 5, 3]])
     """
+    if inplace is not None:
+        warn('The inplace parameter will change'
```
Comment (hmaarrfk) [inline skimage/morphology/_flood_fill.py]:
```suggestion
        warn('The inplace parameter will change '
```

File: skimage/morphology/_flood_fill.py [hunk 3]
```diff
@@ -46,6 +46,11 @@ def flood_fill(image, seed_point, new_value, *, selem=None, connectivity=None,
         If a tolerance is provided, adjacent points with values within plus or
         minus tolerance from the seed point are filled (inclusive).
     inplace : bool, optional
```
Comment (hmaarrfk) [inline skimage/morphology/_flood_fill.py]: Nit: Can we change the order of these two parameters. Have the suggestion one appear first?

File: skimage/morphology/_flood_fill.py [hunk 4]
```diff
@@ -14,7 +14,7 @@
 
 
 def flood_fill(image, seed_point, new_value, *, selem=None, connectivity=None,
-               tolerance=None, inplace=False):
+               tolerance=None, inplace=None, in_place=False):
```
Comment (hmaarrfk) [inline skimage/morphology/_flood_fill.py]:
```suggestion
               tolerance=None, in_place=False, inplace=None):
```

File: skimage/morphology/tests/test_flood_fill.py [hunk 5]
```diff
@@ -112,6 +112,48 @@ def test_inplace_noncontiguous():
...
+    """This test is deprecated and will be removed in
+    version 0.19.0. See #4248.
+    """
+    image = np.array([[0, 0, 0, 0, 0, 0, 0],
+                      [0, 1, 1, 0, 2, 2, 0],
+                      [0, 1, 1, 0, 2, 2, 0],
+                      [1, 0, 0, 0, 0, 0, 3],
+                      [0, 1, 1, 1, 3, 3, 4]])
+
+    flood_fill(image, (0, 0), 5, inplace=True)
```
Comment (hmaarrfk) [inline skimage/morphology/tests/test_flood_fill.py]:
```suggestion
    with expected_warnings(['The inplace']):
        flood_fill(image, (0, 0), 5, inplace=True)
```
Needs a 
```python
from skimage._shared.testing import expected_warnings
```
at the top

File: skimage/morphology/tests/test_flood_fill.py [hunk 6]
```diff
@@ -112,6 +112,48 @@ def test_inplace_noncontiguous():
...
+    """This test is deprecated and will be removed in
+    version 0.19.0. See #4248.
+    """
+    image = np.array([[0, 0, 0, 0, 0, 0, 0],
+                      [0, 1, 1, 0, 2, 2, 0],
+                      [0, 1, 1, 0, 2, 2, 0],
+                      [1, 0, 0, 0, 0, 0, 3],
+                      [0, 1, 1, 1, 3, 3, 4]], dtype=np.float32)
+
+    flood_fill(image, (0, 0), 5, inplace=True)
```
Comment (hmaarrfk) [inline skimage/morphology/tests/test_flood_fill.py]:
```suggestion
    with expected_warnings(['The inplace']):
        flood_fill(image, (0, 0), 5, inplace=True)
```

**Comment by hmaarrfk** [comment]
I requested a few small changes that I think would go a long way to helping maintain the library.

They would mostly hide the warnings in the test suite.

https://dev.azure.com/scikit-image/scikit-image/_build/results?buildId=1733&view=logs&jobId=0c528f75-1409-556f-1999-9f155ac37a3a&taskId=1c50252d-14f6-55f6-616c-dfe500c925e4&lineStart=5675&lineEnd=5676&colStart=1&colEnd=1

Those warnings do help us identify deprecations upstream. Keeping them quiet is useful !

File: skimage/morphology/_flood_fill.py (hunk 2 above)
Comment (lagru) [reply skimage/morphology/_flood_fill.py]: Just a suggestion which may be more clearer:
```
The `inplace` parameter is depreciated and will be removed 
in version 0.19.0. Use `in_place` instead.
```
But I'm fine either.

File: skimage/morphology/_flood_fill.py [hunk 7]
```diff
@@ -100,10 +105,15 @@ def flood_fill(image, seed_point, new_value, *, selem=None, connectivity=None,
            [5, 5, 5, 5, 2, 2, 5],
            [5, 5, 5, 5, 5, 5, 3]])
     """
+    if inplace is not None:
+        warn('The inplace parameter will change'
+             'to in_place in version 0.19.0.', stacklevel=2)
```
Comment (lagru) [inline skimage/morphology/_flood_fill.py]: Could you add `category=FutureWarning` here?

**Comment by rfezzani** [comment]
Isn't this PR related to #4154 and can't we use the same strategy as #4158 for consistency?

**Comment by hmaarrfk** [comment]
You could. But I think it is Overkill and in many cases not as flexible and much more obscure. It also doesn't provide the best stack trace.

Finally, it is really easy for contributor's to understand what is going on without decorators.

I'm fine with both approaches really though.

**Comment by rfezzani** [comment]
Using the decorator have some advantages over the "handmade" deprecation:
- the warning message is standardized
- less lines are introduced to the code base (if multiple such deprecation are introduced like it will be the case according to #4154 )
- the maintenance is easier (only one line to be removed at deprecation cycle end)
- less noise introduced to the documentation, since the deprecated argument is replaced by the new one instead of adding a new argument to the argument list, and only the users concerned by the deprecation are impacted.

**Comment by hmaarrfk** [comment]
I am in full agreement with all the comments above. 

For me, 

> Finally, it is really easy for contributor's to understand what is going on without decorators.

might trump everything ;)

At this stage, there are a few minor fixes, and then this PR can be merged in. Cleanups seem minor, especially seeing that this code won't exist in 1 year.

**Comment by rfezzani** [comment]
I also forgot to mention #4152 as issue to be fixed with the decorator.

File: skimage/morphology/_flood_fill.py [hunk 8]
```diff
@@ -45,10 +45,15 @@ def flood_fill(image, seed_point, new_value, *, selem=None, connectivity=None,
         value of `image` at `seed_point` to be filled.  This is fastest.
         If a tolerance is provided, adjacent points with values within plus or
         minus tolerance from the seed point are filled (inclusive).
-    inplace : bool, optional
+    in_place : bool, optional
         If True, flood filling is applied to `image` inplace.  If False, the
```
Comment (stefanv) [inline skimage/morphology/_flood_fill.py]: Our mistake from before, but note that there is no such English word as "inplace": it should be "in-place".

Comment (robroooh) [reply skimage/morphology/_flood_fill.py]: thank you for the review :+1: this is fixed now.

**Comment by jni** [comment]
Great stuff! Thank you @robroooh and everyone else who reviewed this!

**Pull request merged**